src/libAtomVM/bif.c

```diff
@@ -1512,3 +1515,176 @@ term bif_erlang_size_1(Context *ctx, uint32_t fail_label, int live, term arg1)
+    }
+    long global_atom_index = atom_table_ensure_atom(ctx->global->atom_table, atom, atom_opts);
+    free((void *) atom);
+    if (UNLIKELY(global_atom_index == ATOM_TABLE_NOT_FOUND)) {
+        *error_reason = BADARG_ATOM;
+        return term_invalid_term();
+    } else if (UNLIKELY(global_atom_index == ATOM_TABLE_ALLOC_FAIL)) {
+        *error_reason = OUT_OF_MEMORY_ATOM;
+        return term_invalid_term();
+    }
+    return term_from_atom_index(global_atom_index);
+}
+
+term bif_erlang_binary_to_atom_2(Context *ctx, uint32_t fail_label, int live, term arg1, term arg2)
+{
+    UNUSED(live);
+
+    term error_reason;
+    term result = binary_to_atom(ctx, arg1, arg2, true, &error_reason);
+    if (UNLIKELY(term_is_invalid_term(result))) {
+        RAISE_ERROR_BIF(fail_label, error_reason);
+    }
+    return result;
+}
+
+term bif_erlang_binary_to_existing_atom_2(Context *ctx, uint32_t fail_label, int live, term arg1, term arg2)
+{
+    UNUSED(live);
+
+    term error_reason;
+    term result = binary_to_atom(ctx, arg1, arg2, false, &error_reason);
+    if (UNLIKELY(term_is_invalid_term(result))) {
+        RAISE_ERROR_BIF(fail_label, error_reason);
+    }
+    return result;
+}
+
+term binary_to_atom(Context *ctx, term a_binary, term encoding, bool create_new, term *error_reason)
+{
+    if (UNLIKELY(!term_is_binary(a_binary))) {
+        *error_reason = BADARG_ATOM;
+        return term_invalid_term();
+    }
+
+    const char *atom_string = term_binary_data(a_binary);
+    size_t atom_string_len = term_binary_size(a_binary);
+    if (UNLIKELY(atom_string_len > 255)) {
+        *error_reason = SYSTEM_LIMIT_ATOM;
+        return term_invalid_term();
+    }
+
+    bool encode_latin1_to_utf8 = false;
+    if (UNLIKELY((encoding == LATIN1_ATOM)
+            && !unicode_buf_is_ascii((const uint8_t *) atom_string, atom_string_len))) {
+        encode_latin1_to_utf8 = true;
+    } else if (UNLIKELY((encoding != LATIN1_ATOM) && (encoding != UNICODE_ATOM)
+                   && (encoding != UTF8_ATOM))) {
+        *error_reason = BADARG_ATOM;
+        return term_invalid_term();
+    }
+
+    AtomString atom;
+    if (LIKELY(!encode_latin1_to_utf8)) {
+        size_t i = 0;
+        while (i < atom_string_len) {
+            uint32_t codepoint;
+            size_t codepoint_size;
+            if (UNLIKELY(bitstring_utf8_decode(
+                    (uint8_t *) atom_string + i, atom_string_len, &codepoint, &codepoint_size))
+                != UnicodeTransformDecodeSuccess) {
+                *error_reason = BADARG_ATOM;
+                return term_invalid_term();
+            }
+            i += codepoint_size;
+        }
+
+        atom = malloc(atom_string_len + 1);
+        ((uint8_t *) atom)[0] = atom_string_len;
+        memcpy(((char *) atom) + 1, atom_string, atom_string_len);
+    } else {
+        // * 2 is the worst case size
+        size_t buf_len = atom_string_len * 2;
+        atom = malloc(buf_len + 1);
```

Comment:
No allocation check here either.

Reply:
good catch, this code is old, I will open a PR against release-0.6.

Reply:
Fixed in #1535 for release-0.6, I will fix the conflict when I'll merge release-0.6 into main, after this PR is merged.

Reply:
Good plan. I approved #1535.